File: src/backend/src/XcordTopo.Infrastructure/Validation/TopologyValidator.cs
using System.Text.RegularExpressions;
using XcordTopo.Infrastructure.Providers;
using XcordTopo.Models;

namespace XcordTopo.Infrastructure.Validation;

public sealed partial class TopologyValidator(ProviderRegistry registry) : ITopologyValidator
{
    private static readonly Regex DomainRegex = DomainPattern();
    private static readonly Regex DangerousCharsRegex = DangerousCharsPattern();
    private static readonly HashSet<string> ValidBackupFrequencies =
        new(StringComparer.OrdinalIgnoreCase) { "hourly", "daily", "weekly" };

    public List<string> Validate(Topology topology) =>
        ValidateFull(topology).Errors.Select(e => e.Message).ToList();

    public TopologyValidationResult ValidateFull(Topology topology)
    {
        var items = new List<TopologyValidationError>();

        RunStructuralChecks(topology, items);
        RunDeployErrorChecks(topology, items);
        RunWarningChecks(topology, items);

        return new TopologyValidationResult(items);
    }

    // ─── Tier 0: Existing structural checks ─────────────────────────

    private static void RunStructuralChecks(Topology topology, List<TopologyValidationError> items)
    {
        if (string.IsNullOrWhiteSpace(topology.Name))
            items.Add(new(ValidationSeverity.Error, "Topology name is required.", Field: "name"));

        if (topology.Containers.Count == 0)
            items.Add(new(ValidationSeverity.Error, "Topology must have at least one container."));

        var allPorts = new HashSet<Guid>();
        var allNodeIds = new HashSet<Guid>();
        ValidateContainersRecursive(topology.Containers, items, allPorts, allNodeIds);

        foreach (var wire in topology.Wires)
        {
            if (!allNodeIds.Contains(wire.FromNodeId))
                items.Add(new(ValidationSeverity.Error,
                    $"Wire {wire.Id} references non-existent source node {wire.FromNodeId}."));
            if (!allNodeIds.Contains(wire.ToNodeId))
                items.Add(new(ValidationSeverity.Error,
                    $"Wire {wire.Id} references non-existent target node {wire.ToNodeId}."));
            if (!allPorts.Contains(wire.FromPortId))
                items.Add(new(ValidationSeverity.Error,
                    $"Wire {wire.Id} references non-existent source port {wire.FromPortId}."));
            if (!allPorts.Contains(wire.ToPortId))
                items.Add(new(ValidationSeverity.Error,
                    $"Wire {wire.Id} references non-existent target port {wire.ToPortId}."));
            if (wire.FromNodeId == wire.ToNodeId)
                items.Add(new(ValidationSeverity.Error,
                    $"Wire {wire.Id} cannot connect a node to itself."));
        }

        var wireKeys = new HashSet<string>();
        foreach (var wire in topology.Wires)
        {
            var key = $"{wire.FromPortId}-{wire.ToPortId}";
            var reverseKey = $"{wire.ToPortId}-{wire.FromPortId}";
            if (!wireKeys.Add(key) || wireKeys.Contains(reverseKey))
                items.Add(new(ValidationSeverity.Error,
                    $"Duplicate wire connection between ports {wire.FromPortId} and {wire.ToPortId}."));
        }
    }

    private static void ValidateContainersRecursive(
        List<Container> containers, List<TopologyValidationError> items,
        HashSet<Guid> allPorts, HashSet<Guid> allNodeIds)
    {
        foreach (var container in containers)
        {
            allNodeIds.Add(container.Id);
            foreach (var port in container.Ports)
                allPorts.Add(port.Id);

            foreach (var image in container.Images)
            {
                allNodeIds.Add(image.Id);
                foreach (var port in image.Ports)
                    allPorts.Add(port.Id);

                if (image.Config.TryGetValue("replicas", out var replicas) && !string.IsNullOrEmpty(replicas))
                {
                    if (!IsValidReplicaValue(replicas))
                        items.Add(new(ValidationSeverity.Error,
                            $"Image '{image.Name}' has invalid replicas value '{replicas}'. Must be a positive integer, a range (e.g. 1-3), or a $VARIABLE reference.",
                            NodeId: image.Id.ToString(), Field: "replicas"));
                }
            }

            if (string.IsNullOrWhiteSpace(container.Name))
                items.Add(new(ValidationSeverity.Error,
                    $"Container {container.Id} must have a name.",
                    NodeId: container.Id.ToString(), Field: "name"));

            if (container.Width <= 0 || container.Height <= 0)
                items.Add(new(ValidationSeverity.Error,
                    $"Container '{container.Name}' must have positive dimensions.",
                    NodeId: container.Id.ToString()));

            ValidateContainersRecursive(container.Children, items, allPorts, allNodeIds);
        }
    }

    // ─── Tier 1: Deploy-blocking checks ─────────────────────────────

    private void RunDeployErrorChecks(Topology topology, List<TopologyValidationError> items)
    {
        CheckNameUniqueness(topology, items);
        CheckProviderValidity(topology, items);
        CheckRegionValidity(topology, items);
        CheckTierProfileReferences(topology, items);
        CheckDomainPresence(topology, items);
        CheckVolumeSizes(topology, items);
        CheckComputePoolRequiredImages(topology, items);
        CheckWireCompleteness(topology, items);
        CheckCaddyfileSafety(topology, items);
        CheckHostRamFeasibility(topology, items);
    }

    private static void CheckNameUniqueness(Topology topology, List<TopologyValidationError> items)
    {
        // Container names must be globally unique (they become Terraform resource names)
        var seenContainers = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);

        void Walk(List<Container> containers)
        {
            foreach (var c in containers)
            {
                var sanitized = TopologyHelpers.SanitizeName(c.Name);
                if (!string.IsNullOrEmpty(sanitized))
                {
                    if (seenContainers.TryGetValue(sanitized, out var existing))
                        items.Add(new(ValidationSeverity.Error,
                            $"Name collision: '{c.Name}' and '{existing}' both sanitize to '{sanitized}', causing Terraform resource conflicts.",
                            NodeId: c.Id.ToString(), Field: "name"));
                    else
                        seenContainers[sanitized] = c.Name;
                }

                // Image names only need to be unique within their parent container
                // (they become Docker container names scoped to the host)
                var seenImages = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
                foreach (var img in c.Images)
                {
                    var imgSanitized = TopologyHelpers.SanitizeName(img.Name);
                    if (!string.IsNullOrEmpty(imgSanitized))
                    {
                        if (seenImages.TryGetValue(imgSanitized, out var existing))
                            items.Add(new(ValidationSeverity.Error,
                                $"Name collision: '{img.Name}' and '{existing}' both sanitize to '{imgSanitized}' within '{c.Name}'.",
                                NodeId: img.Id.ToString(), Field: "name"));
                        else
                            seenImages[imgSanitized] = img.Name;
                    }
                }

                Walk(c.Children);
            }
        }

        Walk(topology.Containers);
    }

    private void CheckProviderValidity(Topology topology, List<TopologyValidationError> items)
    {
        var activeKeys = TopologyHelpers.CollectActiveProviderKeys(topology);
        foreach (var key in activeKeys)
        {
            if (registry.Get(key) is null)
                items.Add(new(ValidationSeverity.Error,
                    $"Provider '{key}' is not registered. Valid providers: {string.Join(", ", registry.GetAll().Select(p => p.Key))}"));
        }
    }

    private void CheckRegionValidity(Topology topology, List<TopologyValidationError> items)
    {
        // Check provider-namespaced region keys (e.g., aws_region, linode_region)
        foreach (var provider in registry.GetAll())
        {
            var regionKey = $"{provider.Key}_region";
            if (topology.ProviderConfig.TryGetValue(regionKey, out var region) && !string.IsNullOrEmpty(region))
            {
                var validRegions = provider.GetRegions().Select(r => r.Id).ToHashSet(StringComparer.OrdinalIgnoreCase);
                if (!validRegions.Contains(region))
                    items.Add(new(ValidationSeverity.Error,
                        $"Region '{region}' is not valid for provider '{provider.Key}'. Valid regions: {string.Join(", ", validRegions)}",
                        Field: regionKey));
            }
        }
    }

    private static void CheckTierProfileReferences(Topology topology, List<TopologyValidationError> items)
    {
        var profiles = topology.TierProfiles.Count > 0
            ? topology.TierProfiles
            : ImageOperationalMetadata.DefaultTierProfiles;
        var validIds = profiles.Select(p => p.Id).ToHashSet(StringComparer.OrdinalIgnoreCase);

        void Walk(List<Container> containers)
        {
            foreach (var container in containers)
            {
                if (container.Kind == ContainerKind.ComputePool)
                {
                    var tierRef = container.Config.GetValueOrDefault("tierProfile", "");
                    if (!string.IsNullOrEmpty(tierRef) && !validIds.Contains(tierRef))
                        items.Add(new(ValidationSeverity.Error,
                            $"ComputePool '{container.Name}' references unknown tier profile '{tierRef}'. Valid profiles: {string.Join(", ", validIds)}",
                            NodeId: container.Id.ToString(), Field: "tierProfile"));
                }

                Walk(container.Children);
            }
        }

        Walk(topology.Containers);
    }

    private static void CheckDomainPresence(Topology topology, List<TopologyValidationError> items)
    {
        void Walk(List<Container> containers)
        {
            foreach (var container in containers)
            {
                if (container.Kind is ContainerKind.Caddy or ContainerKind.Dns)
                {
                    var domain = container.Config.GetValueOrDefault("domain", "");
                    if (string.IsNullOrWhiteSpace(domain))
                        items.Add(new(ValidationSeverity.Error,
                            $"{container.Kind} container '{container.Name}' must have a non-empty domain.",
                            NodeId: container.Id.ToString(), Field: "domain"));
                    else if (!DomainRegex.IsMatch(domain))
                        items.Add(new(ValidationSeverity.Error,
                            $"{container.Kind} container '{container.Name}': '{domain}' is not a valid domain name.",
                            NodeId: container.Id.ToString(), Field: "domain"));
                }

                Walk(container.Children);
            }
        }

        Walk(topology.Containers);
    }

    private static void CheckVolumeSizes(Topology topology, List<TopologyValidationError> items)
    {
        void Walk(List<Container> containers)
        {
            foreach (var c in containers)
            {
                foreach (var img in c.Images)
                {
                    if (img.Config.TryGetValue("volumeSize", out var sizeStr) && !string.IsNullOrEmpty(sizeStr))
                    {
                        if (!int.TryParse(sizeStr, out var size) || size <= 0)
                            items.Add(new(ValidationSeverity.Error,
                                $"Image '{img.Name}' in '{c.Name}': volumeSize must be a positive integer, got '{sizeStr}'.",
                                NodeId: img.Id.ToString(), Field: "volumeSize"));
                    }
                }

                Walk(c.Children);
            }
        }

        Walk(topology.Containers);
    }

    private static void CheckComputePoolRequiredImages(Topology topology, List<TopologyValidationError> items)
    {
        // ComputePool no longer requires FederationServer images — they are hub-provisioned at runtime.
        // DataPool should contain at least one data service image.
        void Walk(List<Container> containers)
        {
            foreach (var container in containers)
            {
                if (container.Kind == ContainerKind.DataPool)
                {
                    var hasDataImage = container.Images.Any(i =>
                        i.Kind is ImageKind.PostgreSQL or ImageKind.Redis or ImageKind.MinIO);
                    if (!hasDataImage)
                        items.Add(new(ValidationSeverity.Error,
                            $"DataPool '{container.Name}' must contain at least one data service image (PostgreSQL, Redis, or MinIO).",
                            NodeId: container.Id.ToString()));
                }

                Walk(container.Children);
            }
        }

        Walk(topology.Containers);
    }

    private static void CheckWireCompleteness(Topology topology, List<TopologyValidationError> items)
    {
        var resolver = new WireResolver(topology);

        void Walk(List<Container> containers)
        {
            foreach (var c in containers)
            {
                foreach (var img in c.Images)
                {
                    switch (img.Kind)
                    {
                        case ImageKind.FederationServer:
                            if (resolver.ResolveWiredImage(img.Id, "pg") is null)
                                items.Add(new(ValidationSeverity.Error,
                                    $"FederationServer '{img.Name}' in '{c.Name}' is not connected to a PostgreSQL image.",
                                    NodeId: img.Id.ToString()));
                            if (resolver.ResolveWiredImage(img.Id, "redis") is null)
                                items.Add(new(ValidationSeverity.Error,
                                    $"FederationServer '{img.Name}' in '{c.Name}' is not connected to a Redis image.",
                                    NodeId: img.Id.ToString()));
                            if (resolver.ResolveWiredImage(img.Id, "minio") is null)
                                items.Add(new(ValidationSeverity.Error,
                                    $"FederationServer '{img.Name}' in '{c.Name}' is not connected to a MinIO image.",
                                    NodeId: img.Id.ToString()));
                            break;
                        case ImageKind.HubServer:
                            if (resolver.ResolveWiredImage(img.Id, "pg") is null)
                                items.Add(new(ValidationSeverity.Error,
                                    $"HubServer '{img.Name}' in '{c.Name}' is not connected to a PostgreSQL image.",
                                    NodeId: img.Id.ToString()));
                            if (resolver.ResolveWiredImage(img.Id, "redis") is null)
                                items.Add(new(ValidationSeverity.Error,
                                    $"HubServer '{img.Name}' in '{c.Name}' is not connected to a Redis image.",
                                    NodeId: img.Id.ToString()));
                            break;
                    }
                }

                Walk(c.Children);
            }
        }

        Walk(topology.Containers);
    }

    private static void CheckCaddyfileSafety(Topology topology, List<TopologyValidationError> items)
    {
        void Walk(List<Container> containers)
        {
            foreach (var container in containers)
            {
                if (container.Kind == ContainerKind.Caddy)
                {
                    var domain = container.Config.GetValueOrDefault("domain", "");
                    if (!string.IsNullOrEmpty(domain) && DangerousCharsRegex.IsMatch(domain))
                        items.Add(new(ValidationSeverity.Error,
                            $"Caddy container '{container.Name}': domain value contains unsafe characters that could cause shell injection.",
                            NodeId: container.Id.ToString(), Field: "domain"));
                }

                Walk(container.Children);
            }
        }

        Walk(topology.Containers);
    }

    private void CheckHostRamFeasibility(Topology topology, List<TopologyValidationError> items)
    {
        var hosts = TopologyHelpers.CollectHosts(topology.Containers);
        var provider = registry.Get(topology.Provider);
        if (provider is null) return;

        var plans = provider.GetPlans();
        if (plans.Count == 0) return;
        var maxPlanRam = plans.Max(p => p.MemoryMb);

        foreach (var entry in hosts)
        {
            var required = TopologyHelpers.CalculateHostRam(entry.Host);
            if (required > maxPlanRam)
                items.Add(new(ValidationSeverity.Error,
                    $"Host '{entry.Host.Name}' requires {required} MB RAM but the largest available plan for '{topology.Provider}' is {maxPlanRam} MB.",
                    NodeId: entry.Host.Id.ToString()));
        }
    }

    // ─── Tier 2: Warning checks ─────────────────────────────────────

    private static void RunWarningChecks(Topology topology, List<TopologyValidationError> items)
    {
        CheckOrphanedImages(topology, items);
        CheckUnusedTierProfiles(topology, items);
        CheckBackupFrequency(topology, items);
    }

    private static void CheckOrphanedImages(Topology topology, List<TopologyValidationError> items)
    {
        var connectedNodes = new HashSet<Guid>();
        foreach (var wire in topology.Wires)
        {
            connectedNodes.Add(wire.FromNodeId);
            connectedNodes.Add(wire.ToNodeId);
        }

        void Walk(List<Container> containers)
        {
            foreach (var c in containers)
            {
                foreach (var img in c.Images)
                {
                    if (img.Ports.Count > 0 && !connectedNodes.Contains(img.Id))
                        items.Add(new(ValidationSeverity.Warning,
                            $"Image '{img.Name}' in '{c.Name}' has ports but no wires connecting it.",
                            NodeId: img.Id.ToString()));
                }

                Walk(c.Children);
            }
        }

        Walk(topology.Containers);
    }

    private static void CheckUnusedTierProfiles(Topology topology, List<TopologyValidationError> items)
    {
        if (topology.TierProfiles.Count == 0) return;

        var referencedIds = new HashSet<string>(StringComparer.OrdinalIgnoreCase);

        var hasAgnosticPool = false;

        void Walk(List<Container> containers)
        {
            foreach (var c in containers)
            {
                if (c.Kind == ContainerKind.ComputePool)
                {
                    var tierRef = c.Config.GetValueOrDefault("tierProfile", "");
                    if (!string.IsNullOrEmpty(tierRef))
                        referencedIds.Add(tierRef);
                    else
                        hasAgnosticPool = true; // tier-agnostic pool serves all tiers
                }

                Walk(c.Children);
            }
        }

        Walk(topology.Containers);

        // A tier-agnostic pool (no tierProfile) implicitly references all profiles
        if (hasAgnosticPool) return;

        foreach (var profile in topology.TierProfiles)
        {
            if (!referencedIds.Contains(profile.Id))
                items.Add(new(ValidationSeverity.Warning,
                    $"Tier profile '{profile.Name}' (id: {profile.Id}) is defined but not referenced by any ComputePool."));
        }
    }

    private static void CheckBackupFrequency(Topology topology, List<TopologyValidationError> items)
    {
        void Walk(List<Container> containers)
        {
            foreach (var c in containers)
            {
                var hostFreq = c.Config.GetValueOrDefault("backupFrequency", "");
                if (!string.IsNullOrEmpty(hostFreq) && !ValidBackupFrequencies.Contains(hostFreq))
                    items.Add(new(ValidationSeverity.Warning,
                        $"Container '{c.Name}': backupFrequency '{hostFreq}' is not recognized (valid: hourly, daily, weekly) and will be silently ignored.",
                        NodeId: c.Id.ToString(), Field: "backupFrequency"));

                foreach (var img in c.Images)
                {
                    var imgFreq = img.Config.GetValueOrDefault("backupFrequency", "");
                    if (!string.IsNullOrEmpty(imgFreq) && !ValidBackupFrequencies.Contains(imgFreq))
                        items.Add(new(ValidationSeverity.Warning,
                            $"Image '{img.Name}' in '{c.Name}': backupFrequency '{imgFreq}' is not recognized and will be silently ignored.",
                            NodeId: img.Id.ToString(), Field: "backupFrequency"));
                }

                Walk(c.Children);
            }
        }

        Walk(topology.Containers);
    }

    // ─── Utility ────────────────────────────────────────────────────

    private static bool IsValidReplicaValue(string value)
    {
        if (value.StartsWith('$') && value.Length > 1 && !value.Contains(' '))
            return true;
        if (int.TryParse(value, out var n) && n > 0)
            return true;
        // Accept range format "min-max" where both are positive integers and min <= max
        var parts = value.Split('-');
        if (parts.Length == 2
            && int.TryParse(parts[0], out var min) && min > 0
            && int.TryParse(parts[1], out var max) && max > 0
            && min <= max)
            return true;
        return false;
    }

    [GeneratedRegex(@"^(?:[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]{2,}$")]
    private static partial Regex DomainPattern();

    [GeneratedRegex(@"[;|&$`""\\<>(){}\[\]*?!#]")]
    private static partial Regex DangerousCharsPattern();
}

File: src/frontend/src/lib/topology-validator.ts
import type { Topology, Container, Image, Wire } from '../types/topology';
import type { ValidationItem } from '../types/deploy';

// --- Utility ---

function sanitizeName(name: string): string {
  return name
    .toLowerCase()
    .replace(/ /g, '_')
    .replace(/-/g, '_')
    .replace(/[^a-z0-9_]/g, '');
}

function isValidReplicaValue(value: string): boolean {
  if (value.startsWith('$') && value.length > 1 && !value.includes(' ')) return true;
  const n = parseInt(value, 10);
  if (!isNaN(n) && n > 0 && String(n) === value) return true;
  const parts = value.split('-');
  if (parts.length === 2) {
    const min = parseInt(parts[0], 10);
    const max = parseInt(parts[1], 10);
    if (!isNaN(min) && min > 0 && !isNaN(max) && max > 0 && min <= max
        && String(min) === parts[0] && String(max) === parts[1]) return true;
  }
  return false;
}

const DOMAIN_REGEX = /^(?:[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]{2,}$/;
const DANGEROUS_CHARS_REGEX = /[;|&$`"\\<>(){}\[\]*?!#]/;
const VALID_BACKUP_FREQUENCIES = new Set(['hourly', 'daily', 'weekly']);

// --- Tree walking ---

function walkContainers(containers: Container[], fn: (c: Container) => void): void {
  for (const c of containers) {
    fn(c);
    walkContainers(c.children, fn);
  }
}

function walkImages(containers: Container[], fn: (img: Image, parent: Container) => void): void {
  walkContainers(containers, c => {
    for (const img of c.images) fn(img, c);
  });
}

// --- Wire resolution (simplified client-side) ---

function resolveWiredImage(
  topology: Topology, sourceImageId: string, portName: string
): Image | null {
  // Find the source port by name on the source image
  let sourcePort: { id: string } | undefined;
  walkImages(topology.containers, img => {
    if (img.id === sourceImageId) {
      sourcePort = img.ports.find(p => p.name === portName);
    }
  });
  if (!sourcePort) return null;

  // Find wire from/to this port (wires can be drawn in either direction)
  const wire = topology.wires.find(
    w => w.fromPortId === sourcePort!.id || w.toPortId === sourcePort!.id
  );
  if (!wire) return null;

  // The target node is whichever end ISN'T the source image
  const targetNodeId = wire.fromNodeId === sourceImageId ? wire.toNodeId : wire.fromNodeId;

  // Find target image
  let target: Image | null = null;
  walkImages(topology.containers, img => {
    if (img.id === targetNodeId) target = img;
  });
  return target;
}

// --- Checks ---

function checkStructural(topology: Topology, items: ValidationItem[]): void {
  if (!topology.name?.trim()) {
    items.push({ severity: 'Error', message: 'Topology name is required.', field: 'name' });
  }
  if (topology.containers.length === 0) {
    items.push({ severity: 'Error', message: 'Topology must have at least one container.' });
  }

  const allPorts = new Set<string>();
  const allNodeIds = new Set<string>();

  walkContainers(topology.containers, c => {
    allNodeIds.add(c.id);
    for (const p of c.ports) allPorts.add(p.id);

    if (!c.name?.trim()) {
      items.push({ severity: 'Error', message: `Container must have a name.`, nodeId: c.id, field: 'name' });
    }
    if (c.width <= 0 || c.height <= 0) {
      items.push({ severity: 'Error', message: `Container '${c.name}' must have positive dimensions.`, nodeId: c.id });
    }
  });

  walkImages(topology.containers, (img) => {
    allNodeIds.add(img.id);
    for (const p of img.ports) allPorts.add(p.id);

    const replicas = img.config?.replicas;
    if (replicas && !isValidReplicaValue(replicas)) {
      items.push({
        severity: 'Error',
        message: `Image '${img.name}' has invalid replicas value '${replicas}'. Must be a positive integer, a range (e.g. 1-3), or a $VARIABLE reference.`,
        nodeId: img.id, field: 'replicas',
      });
    }
  });

  // Wire validation
  const wireKeys = new Set<string>();
  for (const wire of topology.wires) {
    if (!allNodeIds.has(wire.fromNodeId))
      items.push({ severity: 'Error', message: `Wire references non-existent source node.` });
    if (!allNodeIds.has(wire.toNodeId))
      items.push({ severity: 'Error', message: `Wire references non-existent target node.` });
    if (!allPorts.has(wire.fromPortId))
      items.push({ severity: 'Error', message: `Wire references non-existent source port.` });
    if (!allPorts.has(wire.toPortId))
      items.push({ severity: 'Error', message: `Wire references non-existent target port.` });
    if (wire.fromNodeId === wire.toNodeId)
      items.push({ severity: 'Error', message: `Wire cannot connect a node to itself.` });

    const key = `${wire.fromPortId}-${wire.toPortId}`;
    const reverseKey = `${wire.toPortId}-${wire.fromPortId}`;
    if (wireKeys.has(key) || wireKeys.has(reverseKey)) {
      items.push({ severity: 'Error', message: `Duplicate wire connection.` });
    }
    wireKeys.add(key);
  }
}

function checkNameUniqueness(topology: Topology, items: ValidationItem[]): void {
  // Container names must be globally unique (they become Terraform resource names)
  const seenContainers = new Map<string, string>();

  walkContainers(topology.containers, c => {
    const sanitized = sanitizeName(c.name);
    if (sanitized) {
      const existing = seenContainers.get(sanitized);
      if (existing !== undefined) {
        items.push({
          severity: 'Error',
          message: `Name collision: '${c.name}' and '${existing}' both sanitize to '${sanitized}', causing Terraform resource conflicts.`,
          nodeId: c.id, field: 'name',
        });
      } else {
        seenContainers.set(sanitized, c.name);
      }
    }

    // Image names only need to be unique within their parent container
    const seenImages = new Map<string, string>();
    for (const img of c.images) {
      const imgSanitized = sanitizeName(img.name);
      if (imgSanitized) {
        const existing = seenImages.get(imgSanitized);
        if (existing !== undefined) {
          items.push({
            severity: 'Error',
            message: `Name collision: '${img.name}' and '${existing}' both sanitize to '${imgSanitized}' within '${c.name}'.`,
            nodeId: img.id, field: 'name',
          });
        } else {
          seenImages.set(imgSanitized, img.name);
        }
      }
    }
  });
}

function checkDomainPresence(topology: Topology, items: ValidationItem[]): void {
  walkContainers(topology.containers, c => {
    if (c.kind === 'Caddy' || c.kind === 'Dns') {
      const domain = c.config?.domain ?? '';
      if (!domain.trim()) {
        items.push({
          severity: 'Error',
          message: `${c.kind} container '${c.name}' must have a non-empty domain.`,
          nodeId: c.id, field: 'domain',
        });
      } else if (!DOMAIN_REGEX.test(domain)) {
        items.push({
          severity: 'Error',
          message: `${c.kind} container '${c.name}': '${domain}' is not a valid domain name.`,
          nodeId: c.id, field: 'domain',
        });
      }
    }
  });
}

function checkVolumeSizes(topology: Topology, items: ValidationItem[]): void {
  walkImages(topology.containers, (img, parent) => {
    const sizeStr = img.config?.volumeSize;
    if (sizeStr) {
      const size = parseInt(sizeStr, 10);
      if (isNaN(size) || size <= 0) {
        items.push({
          severity: 'Error',
          message: `Image '${img.name}' in '${parent.name}': volumeSize must be a positive integer, got '${sizeStr}'.`,
          nodeId: img.id, field: 'volumeSize',
        });
      }
    }
  });
}

function checkComputePoolRequiredImages(topology: Topology, items: ValidationItem[]): void {
  walkContainers(topology.containers, c => {
    if (c.kind === 'ComputePool') {
      const hasFed = c.images.some(i => i.kind === 'FederationServer');
      if (!hasFed) {
        items.push({
          severity: 'Error',
          message: `ComputePool '${c.name}' must contain at least one FederationServer image.`,
          nodeId: c.id,
        });
      }
    }
  });
}

function checkWireCompleteness(topology: Topology, items: ValidationItem[]): void {
  walkImages(topology.containers, (img, parent) => {
    if (img.kind === 'FederationServer') {
      if (!resolveWiredImage(topology, img.id, 'pg'))
        items.push({ severity: 'Error', message: `FederationServer '${img.name}' in '${parent.name}' is not connected to a PostgreSQL image.`, nodeId: img.id });
      if (!resolveWiredImage(topology, img.id, 'redis'))
        items.push({ severity: 'Error', message: `FederationServer '${img.name}' in '${parent.name}' is not connected to a Redis image.`, nodeId: img.id });
      if (!resolveWiredImage(topology, img.id, 'minio'))
        items.push({ severity: 'Error', message: `FederationServer '${img.name}' in '${parent.name}' is not connected to a MinIO image.`, nodeId: img.id });
    }
    if (img.kind === 'HubServer') {
      if (!resolveWiredImage(topology, img.id, 'pg'))
        items.push({ severity: 'Error', message: `HubServer '${img.name}' in '${parent.name}' is not connected to a PostgreSQL image.`, nodeId: img.id });
      if (!resolveWiredImage(topology, img.id, 'redis'))
        items.push({ severity: 'Error', message: `HubServer '${img.name}' in '${parent.name}' is not connected to a Redis image.`, nodeId: img.id });
    }
  });
}

function checkCaddyfileSafety(topology: Topology, items: ValidationItem[]): void {
  walkContainers(topology.containers, c => {
    if (c.kind === 'Caddy') {
      const domain = c.config?.domain ?? '';
      if (domain && DANGEROUS_CHARS_REGEX.test(domain)) {
        items.push({
          severity: 'Error',
          message: `Caddy container '${c.name}': domain value contains unsafe characters.`,
          nodeId: c.id, field: 'domain',
        });
      }
    }
  });
}

function checkOrphanedImages(topology: Topology, items: ValidationItem[]): void {
  const connectedNodes = new Set<string>();
  for (const wire of topology.wires) {
    connectedNodes.add(wire.fromNodeId);
    connectedNodes.add(wire.toNodeId);
  }

  walkImages(topology.containers, (img, parent) => {
    if (img.ports.length > 0 && !connectedNodes.has(img.id)) {
      items.push({
        severity: 'Warning',
        message: `Image '${img.name}' in '${parent.name}' has ports but no wires connecting it.`,
        nodeId: img.id,
      });
    }
  });
}

function checkBackupFrequency(topology: Topology, items: ValidationItem[]): void {
  walkContainers(topology.containers, c => {
    const freq = c.config?.backupFrequency;
    if (freq && !VALID_BACKUP_FREQUENCIES.has(freq.toLowerCase())) {
      items.push({
        severity: 'Warning',
        message: `Container '${c.name}': backupFrequency '${freq}' is not recognized (valid: hourly, daily, weekly).`,
        nodeId: c.id, field: 'backupFrequency',
      });
    }
  });

  walkImages(topology.containers, (img, parent) => {
    const freq = img.config?.backupFrequency;
    if (freq && !VALID_BACKUP_FREQUENCIES.has(freq.toLowerCase())) {
      items.push({
        severity: 'Warning',
        message: `Image '${img.name}' in '${parent.name}': backupFrequency '${freq}' is not recognized.`,
        nodeId: img.id, field: 'backupFrequency',
      });
    }
  });
}

function checkBackupTarget(topology: Topology, items: ValidationItem[]): void {
  const bt = topology.backupTarget;
  if (!bt) return;

  if (!bt.bucketName?.trim()) {
    items.push({
      severity: 'Error',
      message: 'Backup target: bucket name is required.',
      field: 'backupTarget.bucketName',
    });
  }
  if (!bt.region?.trim()) {
    items.push({
      severity: 'Error',
      message: 'Backup target: region is required.',
      field: 'backupTarget.region',
    });
  }
  if (bt.kind === 'S3Compatible' && !bt.endpoint?.trim()) {
    items.push({
      severity: 'Error',
      message: 'Backup target: endpoint is required for S3 Compatible storage.',
      field: 'backupTarget.endpoint',
    });
  }
  if (bt.glacierTransitionDays !== undefined) {
    if (bt.kind !== 'AwsS3') {
      items.push({
        severity: 'Warning',
        message: 'Backup target: glacierTransitionDays is only applicable for AWS S3.',
        field: 'backupTarget.glacierTransitionDays',
      });
    } else if (bt.glacierTransitionDays <= 0) {
      items.push({
        severity: 'Error',
        message: 'Backup target: glacierTransitionDays must be greater than 0.',
        field: 'backupTarget.glacierTransitionDays',
      });
    }
  }
}

// --- Public API ---

export function validateTopology(topology: Topology): ValidationItem[] {
  const items: ValidationItem[] = [];

  // Structural
  checkStructural(topology, items);

  // Deploy-blocking (client-side subset)
  checkNameUniqueness(topology, items);
  checkDomainPresence(topology, items);
  checkVolumeSizes(topology, items);
  checkComputePoolRequiredImages(topology, items);
  checkWireCompleteness(topology, items);
  checkCaddyfileSafety(topology, items);

  // Warnings
  checkOrphanedImages(topology, items);
  checkBackupFrequency(topology, items);

  // Topology-level
  checkBackupTarget(topology, items);

  return items;
}

/** Get all node IDs that have errors or warnings. */
export function getNodeErrors(items: ValidationItem[]): Map<string, ValidationItem[]> {
  const map = new Map<string, ValidationItem[]>();
  for (const item of items) {
    if (item.nodeId) {
      const existing = map.get(item.nodeId);
      if (existing) existing.push(item);
      else map.set(item.nodeId, [item]);
    }
  }
  return map;
}
